### PR TITLE
feat: Workshop Script Editor

### PR DESCRIPTION
**Pull request by Mitcheljager**
This PR adds a web based script editor that allows you to create Workshop scripts using regular text editor functionality. 
This is the first version which is still lacking in a lot of functionality, but it does have:
- Create scripts using syntax highlighting and autocomplete
- Import existing scripts which are parsed to individual items
- Separate everything in to separate "files" and collapsible folders
- Integrated wiki search
- Automatic variable detection

There's a new model, Projects, which essentially holds a big old json object with all the info the editor needs. Each item is separate and doesn't necessarily hold anything specific. It be be 1 rule or 20 rules, it doesn't matter how you separate it. Each item holds it's own content and the position, which is simply 0 to x amount of items and is only used to order the items in the editor. They can be re-ordered by drag and dropping.

The autocomplete is generated from our values and action yml files.

The syntax highlighting is partially dynamic but mostly just a giant dump of keywords from values and actions. This is done through a custom .grammar file which is parsed by lezer to a .lang file. I don't understand how it works, but it does.

The compiler copies the content straight to your clipboard.

The wiki search is a call to our wiki api.

## Editor
image

## Syntax Highlighting
image
image

## Autocomplete
image

## New user experience
image

## Script importer
image

### CODE REVIEW

File: app/javascript/src/components/editor/Empty.svelte [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<div />
+<div class="p-1/2 editor__scrollable">
+  <h1 class="mb-0">Welcome to the Workshop.codes Overwatch Workshop Script Editor</h1>
+  <small class="text-dark">Catchy name, right?</small>
+
+  <div class="line-height">
+    <p>With this editor you can create and edit your Workshop modes using a text based editor. If you have not yet used the Workshop using this editor as a starting point might be difficult. If you have never used a text based editor before but you have used the Workshop, consider this the second step to becoming a programmer!</p>
```
Comment (CactusPuppy) [inline app/javascript/src/components/editor/Empty.svelte:7]: Some small wording and grammar suggestions:
```suggestion
    <p>With this editor, you can create and edit your Workshop modes using a text based editor. If you have not used the Workshop before, using this editor as a starting point might be difficult. If you have never used a text based editor before but you have used the Workshop, consider this the second step to becoming a programmer!</p>
```

File: app/javascript/src/components/editor/Empty.svelte [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        <li>Relevant syntax highlighting.</li>
+        <li>Folders, wow!</li>
+        <li>Integrated Wiki search.</li>
+      </ul>
+    <p>
+
+    <p>Create a new project to get started. Import your existing script or start from zero.</p>
+
+    <p class="text-white">
+      <mark>This editor is currently in an early alpha version. Expect bugs, many of them!</mark>
```
Comment (CactusPuppy) [inline app/javascript/src/components/editor/Empty.svelte:23]:
```suggestion
      <mark>This editor is currently in early alpha. Expect bugs, many of them!</mark>
```

File: app/javascript/src/components/editor/Save.svelte [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    {#if loading}
+      Saving...
+    {:else}
+      Save
+    {/if}
+  </button>
+
+  {#if confettiActive}
+    <div class="confetti">
+      <Confetti colorArray={["var(--primary)"]} x={[-0.5, 0.5]} y={[-0.15, 0.25]} fallDistance="10px" amount="20" size="7" duration=1000 />
```
Comment (CactusPuppy) [inline app/javascript/src/components/editor/Save.svelte:51]: Your Honor, let the record show that ow2countdown use the confetti first /s

**Review by CactusPuppy — CHANGES_REQUESTED** [review]
Overall, a stellar addition to the site. Left some blocking comments, including one security vulnerability.

File: app/controllers/wiki/search_controller.rb [hunk 4]
```diff
@@ -23,7 +23,11 @@ def index
       format.html
       format.json {
         @articles.each do |article|
-          article.content = ReverseMarkdown.convert(ActionController::Base.helpers.sanitize(markdown(article.content || ""), tags: %w(style p br strong em b blockquote h1 h2 h3 h4 h5 h6 code pre)).gsub(/h\d/, "strong"))
+          if params[:parse_markdown]
+            article.content = markdown(article.content)
```
[the remark this answers is not in the document]
Comment (Mitcheljager) [reply app/controllers/wiki/search_controller.rb:27]: Ahh great catch!

File: app/javascript/packs/logged-in-user.js [hunk 5]
```diff
@@ -9,9 +9,10 @@ import SnippetForm from "../src/components/form/Snippet.svelte"
 import TagsForm from "../src/components/form/Tags.svelte"
 import DerivativesForm from "../src/components/form/Derivatives.svelte"
 import Notifications from "../src/components/Notifications.svelte"
+import Editor from "../src/components/editor/Editor.svelte"
 import { LinkedChart, LinkedLabel } from "svelte-tiny-linked-charts"
 
-WebpackerSvelte.setup({ Dropzone, ControlsForm, SnippetForm, TagsForm, DerivativesForm, Notifications, LinkedChart, LinkedLabel })
+WebpackerSvelte.setup({ Dropzone, ControlsForm, SnippetForm, TagsForm, DerivativesForm, Notifications, LinkedChart, LinkedLabel, Editor })
```
[the remark this answers is not in the document]
Comment (Mitcheljager) [reply app/javascript/packs/logged-in-user.js:15]: Good point! I wanted to do this originally but forgot about it

File: app/controllers/wiki/articles_controller.rb [hunk 6]
```diff
@@ -30,9 +30,16 @@ def show
     @edit_ids = Wiki::Article.joins(:edit).approved.where(group_id: @article.group_id).pluck(:"wiki_edits.id")
     @edit_count = Wiki::Edit.where(id: @edit_ids).size
 
+    @article.content = markdown(@article.content) if params[:parse_markdown]
```
[the remark this answers is not in the document]
Comment (Mitcheljager) [reply app/controllers/wiki/articles_controller.rb:33]: Interesting, never thought of that!

**Pull request merged**